todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.kt

```diff
@@ -1,108 +1,111 @@
-
 
 /**
  * Listens to user actions from the list item in ([TasksFragment]) and redirects them to the
  * Fragment's actions listener.
  */
-class TaskDetailViewModel(
-        context: Application,
-        private val tasksRepository: TasksRepository
-) : AndroidViewModel(context), TasksDataSource.GetTaskCallback {
-
-    val task = ObservableField<Task>()
-    val completed = ObservableBoolean()
-    val editTaskCommand = SingleLiveEvent<Void>()
-    val deleteTaskCommand = SingleLiveEvent<Void>()
-    val snackbarMessage = SingleLiveEvent<Int>()
-    var isDataLoading = false
-        private set
-    val isDataAvailable
-        get() = task.get() != null
+open class TaskDetailViewModel(context: Application, private val tasksRepository: TasksRepository) :
+    AndroidViewModel(context), TasksDataSource.GetTaskCallback {
+
+    private val _task = MutableLiveData<Task>()
+
+    private val _isDataAvailable = MutableLiveData<Boolean>()
+
+    private val _dataLoading = MutableLiveData<Boolean>()
+
+    val editTaskCommand = MutableLiveData<Event<Any>>()
+
+    val deleteTaskCommand = MutableLiveData<Event<Any>>()
+
+    private val _snackbarText = MutableLiveData<Event<Int>>()
+
+    // This LiveData depends on another so we can use a transformation.
+    val completed: LiveData<Boolean> = Transformations.map(_task) {
+            input -> input.isCompleted
+    }
+
+    val snackbarMessage: LiveData<Event<Int>>
+        get() = _snackbarText
+
+    val task: LiveData<Task>
+        get() = _task
+
+    val isDataAvailable: LiveData<Boolean>
+        get() = _isDataAvailable
+
+    val dataLoading: LiveData<Boolean>
+        get() = _dataLoading
+
+    protected val taskId: String?
+        get() = _task.value!!.id
 
     fun deleteTask() {
-        task.get()?.let {
-            tasksRepository.deleteTask(it.id)
-            deleteTaskCommand.call()
+        if (_task.value != null) {
+            tasksRepository.deleteTask(_task.value!!.id)
+            deleteTaskCommand.value = Event(Any())
         }
     }
 
     fun editTask() {
-        editTaskCommand.call()
+        editTaskCommand.value = Event(Any())
     }
 
     fun setCompleted(completed: Boolean) {
-        if (isDataLoading) {
+        if (_dataLoading.value!!) {
             return
         }
-        val task = this.task.get().apply {
-            isCompleted = completed
-        }
+        val task = this._task.value
         if (completed) {
-            tasksRepository.completeTask(task)
+            tasksRepository.completeTask(task!!)
             showSnackbarMessage(R.string.task_marked_complete)
         } else {
-            tasksRepository.activateTask(task)
+            tasksRepository.activateTask(task!!)
             showSnackbarMessage(R.string.task_marked_active)
         }
     }
 
     fun start(taskId: String?) {
-        taskId?.let {
-            isDataLoading = true
-            tasksRepository.getTask(it, this)
+        if (taskId != null) {
```

Comment:
`taskId?.apply { `?

Reply:
this was unresolved without change or explanation. so re-opening it.

Reply:
If you use ?.let here (not apply!) `taskId` is not smartly cast to non-null.